static/material/guias/hash.md

```diff
@@ -0,0 +1,268 @@
+```
+
+_Aclaración_: Hay quienes deciden implementar el Hash Abierto de tal forma que no tenga listas vacías (si un lugar aún
+no ha sido utilizado, entonces tener `NULL` y crear la lista cuando sea necesaria). Esto es totalmente válido, y en todo
+caso con aclararlo en algún lado es suficiente (pero debe hacerse la validación contra `NULL` en ese caso).
+
+#### Función
+
+En este caso, no sólo podemos utilizar el iterador externo del hash, sino que **no nos queda otra opción**, dado que no es posible acceder a los campos internos de las estructuras. Además, en particular no nos dicen cuál es la implementación, cosa que no debe ser necesario conocer.
+
+Entonces, simplemente iteramos utilizando el iterador externo y guardamos en una lista. 
+
+```cpp
+lista_t* hash_claves(const hash_t* hash) {
+   lista_t* claves = lista_crear();
+    if (claves) {
+        return NULL;
+    }
+    hash_iter_t* iter = hash_iter_crear(iter);
+    if (!iter) {
+        lista_destruir(claves);
+        return NULL;
+    }
+    for (; ! hash_iter_al_final(iter); hash_iter_avanzar(iter)) {
+        lista_insertar_ultimo(claves, hash_iter_ver_actual(iter));
+    }
+    hash_iter_destruir(iter);
+    return claves;
+}
+```
+
+## Ejercicios propuestos
+
+1.  (★) Suponer que se tiene un hash cerrado que se redimensionar cuando el factor de carga llega a 0.75, y que
+    no se tienen en cuenta los borrados para el factor de carga. Indicar cuál es el peor escenario posible para 
+    esta implementación. 
+
+1.  (★) ¿Para cuáles casos la función `hash_obtener` tiene una complejidad peor que $$\mathcal{O}(1)$$?
+
+1.  (★) Justificar si la siguiente función de hashing es correcta o no: 
+    ```cpp
+    size_t calcular_hash(char *clave, size_t largo) {
+        // rand() devuelve un numero entero positivo aleatorio
+        return rand() % largo;
+    }
+    ```
+
+1.  (★★) a. Mostrar el resultado de las siguientes operaciones tanto para un hash cerrado como para un hash abierto, 
+    ambos de largo 9 (los números son también el resultado de la función de hashing): insertar 17, insertar 22, 
+    insertar 35, borrar 17, insertar 52, insertar 54. 
+
+    b. ¿Qué pasos hay que seguir para verificar si el 70 pertenece al hash? 
+
+    c. Posteriormente se realizan más inserciones. ¿Cuándo redimensionaría cada hash? 
+    ¿Qué pasos hay que seguir para hacerlo?
+
+1.  (★★) Implementar una función de orden $$\mathcal{O}(n)$$ que dado un arreglo de n números enteros devuelva `true` 
+    o `false` según si existe algún elemento que aparezca más de la mitad de las veces. Justificar el orden de 
+    la solución. Ejemplos: 
+
+        [1, 2, 1, 2, 3] -> false
+        [1, 1, 2, 3] -> false 
+        [1, 2, 3, 1, 1, 1] -> true
+        [1] -> true
+
+1.  (★★) Asumiendo que se tiene disponible una implementación completa de Hash, se desea implementar una función 
+    que decida, para un par de tablas de hash, si representan o no el mismo Diccionario. Considere para la solución que 
```

Comment:
>tablas de hash

¿Qué serían “tablas de hash”?  ¿Un _struct hash_cerrado_?

Reply:
Dos TDAs Tabla de Hash. No se sabe como está implementado por dentro.

Reply:
Clarified slightly in f3ef193.